ledger.go

```diff
@@ -200,29 +210,122 @@ func (l *Ledger) Addresses(hrp string, addresses int) ([]*Address, error) {
 	if err != nil {
 		return nil, err
 	}
 	data := []byte{byte(len(addresses))}
 	data = append(data, hash...)
 	data = append(data, pathBytes...)
-	msgHash = append(msgHash, byte(len(data)))
-	msgHash = append(msgHash, data...)
+	data = append([]byte{byte(len(data))}, data...)
 	fmt.Printf("signing hash: %X\n", hash)
-	resp, err := l.device.Exchange(msgHash)
+	resp, err := l.SendToLedger(CLA, INSSignHash, 0x00, 0x00, data)
 	if err != nil {
 		return nil, err
 	}
 	if !bytes.Equal(resp, hash) {
 		return nil, fmt.Errorf("returned hash %x does not match requested %x", resp, hash)
 	}
 
-	return l.collectSignatures(addresses)
+	return l.collectSignatures(addresses, INSSignHash, 0x01, 0x81)
+}
+
+func (l *Ledger) PrepareChunks(txn []byte, addresses []uint32, changePath []uint32) ([][]byte, error) {
+	if txn == nil || addresses == nil {
+		return nil, fmt.Errorf("chunks structure is undefined")
+	}
+	var chunks [][]byte
+	var preamble []byte
+	chunkNum := 0
+
+	pathBytes, err := bip32bytes(pathPrefix, 3)
+	if err != nil {
+		return nil, err
+	}
+	preamble = append(preamble, byte(len(addresses)))
+	preamble = append(preamble, pathBytes...)
+	if changePath != nil {
+		changeBytes, err := bip32bytes(changePath, 3)
+		if err != nil {
+			return nil, err
+		}
+		preamble = append(preamble, changeBytes...)
+	}
+	preamble = append([]byte{byte(len(preamble))}, preamble...)
+	chunks = append(chunks, preamble)
+	chunkNum++
+
+	fmt.Printf("Preamble: %x\n", chunks[0])
+
+	remainingData := txn
+	size := MaxApduSize
+
+	for len(remainingData) > 0 {
+		if len(remainingData) < MaxApduSize {
+			size = len(remainingData)
+		}
+		thisChunk := remainingData[0:size]
+		remainingData = remainingData[size:]
+
+		temp := make([]byte, 0)
+		temp = append(temp, byte(len(thisChunk)))
+		temp = append(temp, thisChunk...)
+		chunks = append(chunks, temp)
+		chunkNum++
+	}
+
+	return chunks, nil
+}
+
+// SignTransaction attempts to sign a valid tx [txn], given a path [addresses]
+// This function will return a signed hash of txn and signatures or an error
+func (l *Ledger) SignTransaction(txn []byte, addresses []uint32, changePath []uint32) ([]byte, [][]byte, error) {
+	if txn == nil || addresses == nil {
+		return nil, nil, fmt.Errorf("transaction or addresses was null")
+	}
+
+	var response []byte
+
+	chunks, err := l.PrepareChunks(txn, addresses, changePath)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	chunkIdx := 0
+
+	// send preamble first
+	preResp, err := l.SendToLedger(CLA, INSSignTransaction, 0x00, 0x00, chunks[chunkIdx])
+	if err != nil {
+		return nil, nil, err
+	}
+	chunkIdx++
+
+	fmt.Printf("Preamble response: %x\n", preResp)
+
+	for chunkIdx < len(chunks)-1 {
+		_, err = l.SendToLedger(CLA, INSSignTransaction, 0x01, 0x00, chunks[chunkIdx])
+		if err != nil {
+			return nil, nil, err
+		}
+		chunkIdx++
+	}
+
+	response, err = l.SendToLedger(CLA, INSSignTransaction, 0x81, 0x00, chunks[chunkIdx])
+	if err != nil {
+		return nil, nil, err
+	}
```

Comment:
I'd personally find this cleaner:
```suggestion
	// send preamble first
	preResp, err := l.SendToLedger(CLA, INSSignTransaction, 0x00, 0x00, chunks[0])
	if err != nil {
		return nil, nil, err
	}

	fmt.Printf("Preamble response: %x\n", preResp)

	lastChunkIndex := len(chunks) - 1
	for i := 1; i < lastChunkIndex; i++ {
		_, err = l.SendToLedger(CLA, INSSignTransaction, 0x01, 0x00, chunks[i])
		if err != nil {
			return nil, nil, err
		}
	}

	response, err = l.SendToLedger(CLA, INSSignTransaction, 0x81, 0x00, chunks[lastChunkIndex])
	if err != nil {
		return nil, nil, err
	}
```